### PR TITLE
Modify run_batch_async

### PR DESCRIPTION
**Pull request by senecameeks**
call run batch async asynchronously

This change gives a 2x speedup when using`run_batch`

### CODE REVIEW

File: cirq-core/cirq/work/sampler.py [hunk 1]
```diff
@@ -293,11 +294,18 @@ async def run_batch_async(
 
         See docs for `cirq.Sampler.run_batch`.
         """
+        print("finally")
         params_list, repetitions = self._normalize_batch_args(programs, params_list, repetitions)
-        return [
-            await self.run_sweep_async(circuit, params=params, repetitions=repetitions)
-            for circuit, params, repetitions in zip(programs, params_list, repetitions)
-        ]
```
Comment (maffoo) [inline cirq-core/cirq/work/sampler.py:300]: You probably want to use a variant of `duet.pmap` ("parallel map") to make multiple calls in parallel. In this case, the `zip` produces tuples of args and we want to call the underlying function like `self.run_sweep_async(*args)`, so we need a "starmap", something like:
```python
return await duet.pstarmap_async(self.run_sweep_async, zip(programs, params_list, repetitions))
```

Comment (maffoo) [reply cirq-core/cirq/work/sampler.py:300]: For reference, here's the definition of `duet.pstarmap_async`: https://github.com/google/duet/blob/main/duet/api.py#L171

Comment (senecameeks) [reply cirq-core/cirq/work/sampler.py:300]: Thank you for the code pointer!

File: cirq-core/cirq/work/sampler.py [hunk 2]
```diff
@@ -294,10 +294,9 @@ async def run_batch_async(
         See docs for `cirq.Sampler.run_batch`.
         """
         params_list, repetitions = self._normalize_batch_args(programs, params_list, repetitions)
-        return [
-            await self.run_sweep_async(circuit, params=params, repetitions=repetitions)
-            for circuit, params, repetitions in zip(programs, params_list, repetitions)
-        ]
+        return await duet.pstarmap_async(
```
Comment (wcourtney) [inline cirq-core/cirq/work/sampler.py:297]: Can you add a test that demonstrates the list of programs are all sent without waiting on any to complete? (This should fail under the former implementation). Thanks!

**Review by wcourtney — APPROVED** [review]
LGTM % some test nits. Thanks!

**Review by mpharrigan — APPROVED** [review]

**Pull request merged**